homeassistant/components/ultrasync/__init__.py

```diff
@@ -0,0 +1,136 @@
+
+PLATFORMS = ["sensor"]
+
+
+async def async_setup(hass: HomeAssistantType, config: dict) -> bool:
+    """Set up the UltraSync integration."""
+    hass.data.setdefault(DOMAIN, {})
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry) -> bool:
+    """Set up UltraSync from a config entry."""
+    if not entry.options:
+        options = {
+            CONF_SCAN_INTERVAL: entry.data.get(
+                CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL
+            ),
+        }
+        hass.config_entries.async_update_entry(entry, options=options)
+
+    coordinator = UltraSyncDataUpdateCoordinator(
+        hass,
+        config=entry.data,
+        options=entry.options,
+    )
+
+    await coordinator.async_refresh()
+
+    if not coordinator.last_update_success:
+        raise ConfigEntryNotReady
+
+    undo_listener = entry.add_update_listener(_async_update_listener)
+
+    hass.data[DOMAIN][entry.entry_id] = {
+        DATA_COORDINATOR: coordinator,
+        DATA_UNDO_UPDATE_LISTENER: [undo_listener],
+        SENSORS: {},
+    }
+
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    _async_register_services(hass, coordinator)
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistantType, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
+
+    if unload_ok:
+        for unsub in hass.data[DOMAIN][entry.entry_id][DATA_UNDO_UPDATE_LISTENER]:
+            unsub()
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    return unload_ok
+
+
+def _async_register_services(
+    hass: HomeAssistantType,
+    coordinator: UltraSyncDataUpdateCoordinator,
+) -> None:
+    """Register integration-level services."""
+
+    def away(call) -> None:
+        """Service call to set alarm system to 'away' mode in UltraSync Hub."""
+        coordinator.hub.set_alarm(state=AlarmScene.AWAY)
+
+    def stay(call) -> None:
+        """Service call to set alarm system to 'stay' mode in UltraSync Hub."""
+        coordinator.hub.set_alarm(state=AlarmScene.STAY)
+
+    def disarm(call) -> None:
+        """Service call to disable alarm in UltraSync Hub."""
+        coordinator.hub.set_alarm(state=AlarmScene.DISARMED)
+
+    hass.services.async_register(DOMAIN, SERVICE_AWAY, away, schema=vol.Schema({}))
+    hass.services.async_register(DOMAIN, SERVICE_STAY, stay, schema=vol.Schema({}))
+    hass.services.async_register(DOMAIN, SERVICE_DISARM, disarm, schema=vol.Schema({}))
+
+
+async def _async_update_listener(hass: HomeAssistantType, entry: ConfigEntry) -> None:
+    """Handle options update."""
+    await hass.config_entries.async_reload(entry.entry_id)
```

Comment:
Reloading the config entries should not be necessary, since you are using a coordinator. You could just set the new update interval directly on the coordinator.

```python
    if entry.options[CONF_UPDATE_INTERVAL]:
        coordinator = hass.data[DOMAIN][entry.entry_id]["coordinator"]
        coordinator.update_interval = timedelta(seconds=entry.options[CONF_UPDATE_INTERVAL])

        await coordinator.async_refresh()
```

Reply:
I put this code in place; i'll give it a whirl